Add support for k8s 1.21

aanm opened the pull request:
```release-note
Add support for k8s 1.21 and set minimal k8s supported version to 1.16
```

read per commit basis

@brb PTAL at `test/provision/k8s_install.sh` changes.

aanm committed:
docs: bump minimal supported k8s version to 1.16

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
add support for k8s 1.21.0

This commit adds the minimal set of changes required to be able to
compile Cilium with the new k8s libraries as well as it synced all files
in pkg/k8s/slim with the upstream versions.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: add tests for k8s 1.21 and set min k8s version to 1.16

Signed-off-by: André Martins <andre@cilium.io>

christarazi reviewed (approved):
:rocket:

nebril commented on Documentation/concepts/kubernetes/compatibility.rst:
```diff
@@ -17,12 +17,12 @@ with Cilium. Older Kubernetes versions not listed in this table do not have
+| k8s Version                        | k8s NetworkPolicy API     | CiliumNetworkPolicy        |
++------------------------------------+---------------------------+----------------------------+
+|                                    |                           | ``cilium.io/v2`` has a     |
+| 1.16, 1.17, 1.18, 1.19, 1.20, 1.21 | * `networking.k8s.io/v1`_ | `CustomResourceDefinition` |
```
Removing 3 versions while adding only 1? Oh yiss.

nebril commented on jenkinsfiles/ginkgo-kubernetes-all.Jenkinsfile:
```diff
@@ -129,7 +129,7 @@ pipeline {
             }
         }
-        stage('Copy code and boot VMs 1.{14,15}'){
+        stage('Copy code and boot VMs 1.{17,18}'){
```
@nbusseneau I think it's pointless to change this file, should we delete it on master?

nbusseneau replied:
Yes, it is not used anymore on `master`, can be removed. We kept it while transitioning quarantine builds but this has been done since then.

nebril commented on jenkinsfiles/ginkgo.Jenkinsfile:
```diff
@@ -125,13 +125,13 @@ pipeline {
                         }
                     }
                 }
-                stage('Boot vms K8s-1.13 net-next kubeproxy-free') {
```
This file also doesn't get run on master. @nbusseneau also a candidate for deletion?

nbusseneau replied:
It is still ran from obsolete job https://jenkins.cilium.io/view/Obsolete/job/cilium-ginkgo/job/cilium/job/master/, but nobody looks at that it seems.

nbusseneau replied:
Other than that I think it can safely be removed on `master`, yes.

brb commented on test/provision/k8s_install.sh:
```diff
@@ -315,13 +315,25 @@ case $K8S_VERSION in
-#if [ "$KUBEPROXY" == "0" ]; then
-#    KUBEADM_OPTIONS="$KUBEADM_OPTIONS --skip-phases=addon/kube-proxy"
-#fi
+if [ "$KUBEPROXY" == "0" ]; then
```
Great! You can now remove:
```
if [[ "${KUBEPROXY}" == "0" ]]; then
          kubectl -n kube-system delete ds kube-proxy
          iptables-restore <(iptables-save | grep -v KUBE)
      fi
```

nebril reviewed (approved):
My comments are non blocking. We probably need to delete these obsolete jenkinsfiles, but it's out of scope of this PR.

tklauser commented on contrib/vagrant/deployments/coredns-controller.yaml:
```diff
@@ -52,7 +52,7 @@ spec:
       containers:
       - name: coredns
-        image: k8s.gcr.io/coredns:1.7.0
+        image: k8s.gcr.io/coredns/coredns:v1.8.0
```
Probably need to update https://github.com/cilium/packer-ci-build/blob/master/provision/pull-images.sh accordingly.

aanm replied:
Done https://github.com/cilium/packer-ci-build/pull/260

tklauser reviewed (approved):
Vendoring LGTM.

qmonnet reviewed (changes requested):
Grepping on the sources, I can find several locations that still refer to v1.13, in the code or in comments (a few of them also missing a reference to v1.21). Can you please double check? In particular:

- [ ] pkg/k8s/cnp_tests.go
      Irrelevant
- [ ] install/kubernetes/cilium/values.yaml (but in a comment only)
      I will leave this one untouched and remove it in the next release
- [x] install/kubernetes/cilium/Chart.yaml
- [x] install/kubernetes/cilium/README.md
- [x] install/kubernetes/Makefile (comment)
- [x] test/provision/k8s_install.sh
- [x] test/k8sT/istio.go
- [x] test/k8sT/Policies.go
- [x] Documentation/contributing/testing/e2e.rst
- [x] Documentation/gettingstarted/requirements_intro.rst
- [x] Documentation/gettingstarted/istio.rst

Additionally, I found references to v1.14 in those files:

- [ ] pkg/k8s/synced/crd.go
     @christarazi can you take a look at this one?
- [ ] pkg/k8s/version/version.go
      I will leave this one untouched and remove it in the next release
- [ ] install/kubernetes/experimental-install.yaml
      @qmonnet this file doesn't exist in the tree ;-)
- [x] install/kubernetes/cilium/templates/cilium-operator-clusterrole.yaml
- [x] install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
- [x] install/kubernetes/cilium/templates/_helpers.tpl
- [x] test/k8sT/Services.go
- [ ] test/packet/README.md
      Irrelevant
- [ ] Documentation/internals/cilium_operator.rst
      Irrelevant

And to v1.15 in:

- [x] Documentation/gettingstarted/kubeproxy-free.rst
- [x] test/k8sT/Services.go (different location than for v1.14)

edited by @aanm

qmonnet reviewed (commented):
Thanks for the changes! The following are marked as addressed but still contain references that may be outdated. Please have a last look, if you think they're irrelevant then I'm good with it.

```
$ rg '1\.1[3-5]' install/kubernetes/cilium/templates/_helpers.tpl test/k8sT/Policies.go test/provision/k8s_install.sh
test/provision/k8s_install.sh
236:    "1.13")
238:        K8S_FULL_VERSION="1.13.12"
244:    "1.14")
246:        K8S_FULL_VERSION="1.14.10"
252:    "1.15")
254:        K8S_FULL_VERSION="1.15.12"

test/k8sT/Policies.go
1824:                   case "1.13", "1.14", "1.15":

install/kubernetes/cilium/templates/_helpers.tpl
12:{{- if semverCompare ">=1.14-0, <1.19-0" .Capabilities.KubeVersion.Version -}}
```

nbusseneau reviewed (approved):
Made my comments in response to Maciej above.

qmonnet reviewed (approved):
Thanks!

nbusseneau reviewed (changes requested):
In `Documentation/contributing/testing/ci.rst`, we should add `test-backport-1.10` to the list of main triggers (line 180).

Unrelated: just discovered that GitHub forbids adding review comments on collapsed lines in a modified file.

christarazi committed:
k8s: Remove references to 1.14

This partially reverts a73cb4d84d ("k8s: Add ability to watch for full
CRD object") as the code handling versions of K8s 1.14 below is no
longer needed.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: André Martins <andre@cilium.io>

christarazi committed:
Revert "k8s: Add capability flag for watching metav1.POM"

This commit is no longer needed as the minimally supported version is
now K8s 1.16.

This reverts commit c9474acb45b06bdecc7c58a2d6fd3e7bf828f008.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
